### ISSUE TITLE
How to set num_training_steps in lr_scheduler properly

### ISSUE THREAD

**Issue by allanj** [huggingface/accelerate#628]
Usually I call something like this to set the scheduler

```python3
from transformers import get_linear_schedule_with_warmup
scheduler = get_linear_schedule_with_warmup(
        optimizer, num_warmup_steps=warmup_step, num_training_steps=num_training_steps
    )
```

And the `num_training_steps` usually equals to
```python3
t_total = int(len(train_dataloader) * num_epochs) ## len(train_dataloader) = num_batches
```
If I use accelerator, should I change the `num_training_steps` to something like this? And how to understand this:

I believe the `len(train_dataloader) = batch size at each device`

```python3
t_total = int(len(train_dataloader) * num_epochs // accelerator.num_processes)
```

All the above operations are before `accelerator.prepare`

**Comment by sgugger** [comment]
No, Accelerate will take into account the number of processes for you, as long as you do send your `lr_scheduler` to the `Accelerator.prepare` method. You can thus use:
```py
t_total = int(len(train_dataloader) * num_epochs) ## len(train_dataloader) = num_batches
```

**Comment by takiholadi** [comment]
then maybe update needed here: https://github.com/huggingface/accelerate/blob/b0f8189d34fa42821ca041e2cba161db864c76b5/examples/complete_nlp_example.py#L141

**Comment by allanj** [comment]
> then maybe update needed here:
> 
> https://github.com/huggingface/accelerate/blob/b0f8189d34fa42821ca041e2cba161db864c76b5/examples/complete_nlp_example.py#L141

I guess that means we don't need to update that

**Comment by muellerzr** [comment]
Yep! Will do so today, thanks for the heads up!

**Comment by allanj** [comment]
Hi @muellerzr , I don't quite understand, I thought the sample sent by @takiholadi  has no `accelerator.num_processed`. Why we should revise that

**Comment by takiholadi** [comment]
maybe also need to update: https://huggingface.co/docs/accelerate/main/en/concept_guides/gradient_synchronization

**Comment by takiholadi** [comment]
> Hi @muellerzr , I don't quite understand, I thought the sample sent by @takiholadi has no `accelerator.num_processed`. Why we should revise that

Since:
https://github.com/huggingface/accelerate/blob/main/src/accelerate/scheduler.py#L31-L32

**Comment by muellerzr** [comment]
@allanj correct, everything and anything that has to do with gradient accumulation accelerate will now handle for you. Just pass in the `gradient_accumulation_steps` arg and make no changes to your code, as if you weren't using gradient accumulation at all 😄

**Comment by allanj** [comment]
Thanks. I was actually also trying that. But somehow run into an issue when I combine this functionality (https://huggingface.co/docs/accelerate/usage_guides/gradient_accumulation) with `clip_grad_norm`

```python3
with torch.cuda.amp.autocast(enabled=bool(args.fp16)), accelerator.accumulate(model):
    output = model(**batch)
    loss = output.loss
    total_loss += loss.item()
    accelerator.backward(loss)
    accelerator.clip_grad_norm_(model.parameters(), args.max_grad_norm)
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
```

I didn't copy the error. But the error happens in the line of gradient clipping, showing something similar to below
```bash
unscale() has already been called...
```

> @allanj everything and anything that has to do with gradient accumulation accelerate will now handle for you. Just pass in the `gradient_accumulation_steps` arg and make no changes to your code, as if you weren't using gradient accumulation at all 😄

**Comment by songlin** [comment]
@allanj  I had the same problem, did you solve it ?

**Comment by allanj** [comment]
Sorry. I don't quite have the impression about how I fixed it.
